### PR TITLE
extmod/modbluetooth: Addressing modes, Unix H4 UART, reliability fixes (v2)

### PR DESCRIPTION
**Pull request by dpgeorge**
This is #6343 with some minor changes:
- support for address rotation removed (it's not working properly yet)
- removed commit that "Automatically notify subscribed connections on gatts_write()." because it fails the BLE multitest (and it's not easy to fix at this stage)

### CODE REVIEW

**Comment by andrewleech** [comment]
Have you got any info/idea of what's wrong with the "notify subscribed connections" change? 
I did always think it seemed wrong to need that command there that way to make notification subscriptions work, but I'm relying on it for my current application and it appears to be working for me?

**Comment by dpgeorge** [comment]
> Have you got any info/idea of what's wrong with the "notify subscribed connections" change?

The commit itself is not necessarily wrong.  It's that it makes the `ble_characteristic.py` multi-test fail and fixing that is hard (there are now extra indicates being sent and one of them fails).

**Comment by andrewleech** [comment]
Ah yes, I see what you mean!

File: ports/stm32/Makefile [hunk 1]
```diff
@@ -490,12 +490,13 @@ endif
 
 ifeq ($(MICROPY_BLUETOOTH_NIMBLE),1)
 include $(TOP)/extmod/nimble/nimble.mk
-SRC_C += nimble.c
+SRC_C += mpnimbleport.c
 endif
 
 ifeq ($(MICROPY_BLUETOOTH_BTSTACK),1)
+MICROPY_BLUETOOTH_BTSTACK_H4 ?= 1
```
Comment (dpgeorge) [inline ports/stm32/Makefile:497]: why is this set here?

File: extmod/btstack/btstack.mk [hunk 2]
```diff
@@ -38,6 +38,18 @@ CFLAGS  += $(shell pkg-config libusb-1.0 --cflags)
 LDFLAGS += $(shell pkg-config libusb-1.0 --libs)
 endif
 
+ifeq ($(MICROPY_BLUETOOTH_BTSTACK_H4),1)
```
Comment (dpgeorge) [inline extmod/btstack/btstack.mk:49]: does this H4 stuff belong in this commit?

File: ports/unix/mpbtstackport_common.c [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+#include "lib/btstack/platform/embedded/hal_time_ms.h"
+
+#include "extmod/btstack/modbluetooth_btstack.h"
+
+// Called by the UART polling thread in mphciport.c, or by the USB polling thread in mpbtstackport_usb.c.
+bool mp_bluetooth_hci_poll(void) {
+    if (mp_bluetooth_btstack_state == MP_BLUETOOTH_BTSTACK_STATE_STARTING || mp_bluetooth_btstack_state == MP_BLUETOOTH_BTSTACK_STATE_ACTIVE || mp_bluetooth_btstack_state == MP_BLUETOOTH_BTSTACK_STATE_HALTING) {
+        // Pretend like we're running in IRQ context (i.e. other things can't be running at the same time).
+        mp_uint_t atomic_state = MICROPY_BEGIN_ATOMIC_SECTION();
+        #if MICROPY_BLUETOOTH_BTSTACK_H4
```
Comment (dpgeorge) [inline ports/unix/mpbtstackport_common.c:46]: does this belong in this commit?

File: ports/unix/mpbtstackport_common.c [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        void mp_bluetooth_hci_poll_h4(void);
```
Comment (dpgeorge) [inline ports/unix/mpbtstackport_common.c:47]: this decl can be removed

File: ports/unix/mpbtstackport_common.c [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        run_loop_init = true;
+        btstack_run_loop_init(btstack_run_loop_embedded_get_instance());
+    } else {
+        btstack_run_loop_embedded_get_instance()->init();
+    }
+
+    // hci_dump_open(NULL, HCI_DUMP_STDOUT);
+
+    #if MICROPY_BLUETOOTH_BTSTACK_USB
+    void mp_bluetooth_btstack_port_init_usb(void);
```
Comment (dpgeorge) [inline ports/unix/mpbtstackport_common.c:92]: this decl can be removed

**Comment by tve** [comment]
The issue with cores on the esp32 comes down to MICROPY_BEGIN_ATOMIC_SECTION being aliases to portENTER_CRITICAL_NESTED which is not multi-core safe. I can't find info on this: does MICROPY_BEGIN_ATOMIC_SECTION have to support nesting? Because if not, MICROPY_PY_BLUETOOTH_ENTER could use portENTER_CRITICAL, which is multi-core safe.

**Comment by jimmo** [comment]
@tve 
I agree that this should only be done if BLE is enabled.

I'd like to fix this (and really get to the bottom of the threading/multicore stuff on ESP32) but haven't had a chance to really get into the details.

> MICROPY_BEGIN_ATOMIC_SECTION being aliases to portENTER_CRITICAL_NESTED which is not multi-core safe

I've noticed this previously, and wondered why this isn't a problem more generally (outside BLE). I've never looked into the multi-threading implementation on ESP32. I'm guessing this works currently (despite not being multi-core-safe) for Python threads because all MicroPython tasks still run on the same core.

> For example, I virtually always use a build with BL disabled so I get both cores and a lot of memory back.

Out of curiosity, what do you mean by getting "both cores back" (given that all MicroPython threads will always be on the same core). I guess there's enough overhead of other ESP32 functionality running that having a dedicated core for all MicroPython threads is still a win.

> does MICROPY_BEGIN_ATOMIC_SECTION have to support nesting?

I know on Unix when implementing the equivalent I had to use a recursive pthread mutex, but that might have been specifically because of the way the BLE code uses the mutex.

But would moving it to use portENTER_CRITICAL actually solve the slow-startup issue that the pinning is working around? I think @dpgeorge said he could reproduce in an IDF example (and it's fixed in later IDF versions?).  (Ref https://github.com/micropython/micropython/issues/5489#issuecomment-617665343  ).

**Comment by tve** [comment]
> I've noticed this previously, and wondered why this isn't a problem more generally (outside BLE).

Because most (everything?) provided by esp-idf is multicode-safe and python is constrained to one core. One other place this comes up is with wifi-based communication. For std wifi (sockets) this is handled in the esp-idf network stack, which is one reason the esp32's modsocket doesn't use low level LwIP APIs 'cause they're not multi-core safe. Another place is ESP-NOW support, whose esp-idf api is not multi-core safe per-se either, which is why Glenn had to add a ring buffer (how many ring buffer implementations do we have, in addition to the one esp-idf/freertos provide?).

> what do you mean by getting "both cores back"

What I mean is that MP can run on a core relatively undisturbed by whatever RF stuff esp-idf does, which does impact latencies.

> But would moving it to use portENTER_CRITICAL actually solve the slow-startup issue that the pinning is working around?

I'm unfortunately not familiar with the bluetooth support. So far I'm having a hell of a time getting a simple BLE client for a heart rate sensor to connect/reconnect reliably after something happens, but to be fair, it could also be on my end.

**Pull request merged**